### PR TITLE
magento/graphql-ce:#678 Send email to friend

### PR DESCRIPTION
**Pull request by rleshchenko**
### Description (*)
Previously there was using comparison with Magento\Catalog\Model\Product\Attribute\Source\Status::STATUS ENABLED instead of 
comparison with Magento\Catalog\Model\Product\Visibility::VISIBILITY*. Please look at the Magento\Catalog\Model\Product::isVisibleInCatalog()

### Fixed Issues (if relevant)

1. magento/magento2#<#678>: Issue title

### Manual testing scenarios (*)

1. ...
2. ...

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/graphql-ce?pullRequest=752) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dev.magento.com/magento/graphql-ce?pullRequest=752) before we can accept your contribution.<br/>

File: app/code/Magento/SendFriendGraphQl/Model/Resolver/SendEmailToFriend.php [hunk 1]
```diff
@@ -45,22 +47,30 @@ class SendEmailToFriend implements ResolverInterface
...
      * @param ManagerInterface $eventManager
+     * @param Visibility|null $visibility
      */
     public function __construct(
         SendFriendFactory $sendFriendFactory,
         ProductRepositoryInterface $productRepository,
         DataObjectFactory $dataObjectFactory,
-        ManagerInterface $eventManager
+        ManagerInterface $eventManager,
+        ?Visibility $visibility= null
```
Comment (vpodorozh) [inline app/code/Magento/SendFriendGraphQl/Model/Resolver/SendEmailToFriend.php]: Backward compatibility should not be followed in GraphQL php codebase - only schema counted as a contract.
https://github.com/magento/graphql-ce/wiki/Approach-to-Implementation#backward-compatibility-policy

File: app/code/Magento/SendFriendGraphQl/Model/Resolver/SendEmailToFriend.php [hunk 2]
```diff
@@ -45,22 +47,30 @@ class SendEmailToFriend implements ResolverInterface
...
     ) {
         $this->sendFriendFactory = $sendFriendFactory;
         $this->productRepository = $productRepository;
         $this->dataObjectFactory = $dataObjectFactory;
         $this->eventManager = $eventManager;
+        $this->visibility = $visibility ?: ObjectManager::getInstance()->create(Visibility::class);
```
Comment (vpodorozh) [inline app/code/Magento/SendFriendGraphQl/Model/Resolver/SendEmailToFriend.php]: check prev comment

**Review by vpodorozh — CHANGES_REQUESTED** [review]
Please check my comments and fix static tests.
Thx!

**Comment by vpodorozh** [comment]
@rleshchenko - please move `\Magento\SendFriendGraphQl\Model\Resolver\SendEmailToFriend::getProduct` into separate service. thx - this will eliminate two dependencies that break tests

**Comment by rleshchenko** [comment]
@vpodorozh, I've refactored code according to your notes. Please look again

File: app/code/Magento/SendFriendGraphQl/Model/Resolver/ProductService.php [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+                    __("The product that was requested doesn't exist. Verify the product and try again.")
+                );
+            }
+        } catch (NoSuchEntityException $e) {
+            throw new GraphQlNoSuchEntityException(__($e->getMessage()), $e);
+        }
+        return $product;
+    }
+
+}
```
Comment (rleshchenko) [inline app/code/Magento/SendFriendGraphQl/Model/Resolver/ProductService.php]:
```suggestion
}

```

File: app/code/Magento/SendFriendGraphQl/Model/Resolver/ProductService.php [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+                throw new GraphQlNoSuchEntityException(
+                    __("The product that was requested doesn't exist. Verify the product and try again.")
+                );
+            }
+        } catch (NoSuchEntityException $e) {
+            throw new GraphQlNoSuchEntityException(__($e->getMessage()), $e);
+        }
+        return $product;
+    }
+
```
Comment (rleshchenko) [inline app/code/Magento/SendFriendGraphQl/Model/Resolver/ProductService.php]:
```suggestion
```

File: app/code/Magento/SendFriendGraphQl/Model/Resolver/ProductService.php [hunk 5]
```diff
@@ -0,0 +1,65 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+declare(strict_types=1);
+
+
```
Comment (rleshchenko) [inline app/code/Magento/SendFriendGraphQl/Model/Resolver/ProductService.php]:
```suggestion
```

File: app/code/Magento/SendFriendGraphQl/Model/Resolver/ProductService.php [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+use Magento\Catalog\Api\ProductRepositoryInterface;
+use Magento\Catalog\Model\Product\Visibility;
+use Magento\Catalog\Api\Data\ProductInterface;
+use Magento\Framework\GraphQl\Exception\GraphQlNoSuchEntityException;
+use Magento\Framework\Exception\NoSuchEntityException;
+
+/**
+ * Class ProductService
+ */
+class ProductService
```
Comment (vpodorozh) [inline app/code/Magento/SendFriendGraphQl/Model/Resolver/ProductService.php]: @rleshchenko:
1. could you please use the `Command` pattern for this.
Like - class name `GetProduct` and only one method is present called `execute`.

2. Please move to `Provider` directory - for example `\Magento\PaypalGraphQl\Model\Provider\Config` (sorry that I did not mention it earlier)

**Review by vpodorozh — CHANGES_REQUESTED** [review]

**Comment by vpodorozh** [comment]
@novikor - please pick up this PR.
Everything seems to be ok from my point of view - so can you just approve it? (from now I do not have permissions any more)

Thx!

**Comment by lenaorobei** [comment]
@rleshchenko thank you for the contribution. Few changes I made to your PR:
 - merged @pmclain `pmclain:issue/732` branch to your PR since you both were fixing `SendFriend` module;
- resolved conflicts;
- fixed expected exception message in the `api-functional test`.

**Review by naydav — APPROVED** [review]

**Pull request merged**